List utilsby with new exports

yanick opened the pull request:
If you are on-board with #2, use this instead of #3.

autarch commented on t/01-basic.t:
```diff
@@ -89,17 +90,6 @@ is_deeply(
 );
 
-is(
-    ( List::AllUtils::first { $_ > 5 } ( 1, 2, 5, 22, 7 ) ),
```
If this is no longer working then that's a problem.

The reason I tested this is that there are other modules, like `Test::Deep`, that export subs with names that overlap with subs from `List::AllUtils`. In that case, you have to call _something_ by it's FQ name. I don't think we should break that feature of `List::AllUtils`. Maybe an `AUTOLOAD` would do the trick?

yanick replied:
Aaaah, so that's why. Makes sense.

I don't think we can use `AUTOLOAD` here, as perl must know the functions' prototype beforehand do it won't panic seeing things like `any { ... } @blah`.  But I have a plan B. There will be an update on this PR somewhen after breakfast. :-)

autarch commented:
I like the concept, but there's an issue I noted. Also please run `tidyall` on the PR if you can. Thanks!

yanick commented:
Everything tidied. 

As for the availability in the List::AllUtils:: namespace, what I did was to create shim functions that will use the relevant module on-demand. Using an AUTOLOAD trick would have been nice, but we need to have the `(&@)`-like prototypes in the first place.

autarch commented:
There are a number of legit failures in travis, including a spelling mistake (possibly copied from List::UtilsBy?) and what looks like a prereq issue. I think the sanest thing to do for this distro is to simply depend on List::Util 1.45 and not worry about earlier versions. Is there any reason to go to great lengths to make this work with earlier versions of List::Util? I did that in the existing version simply because it's so easy, but with your changes, it becomes much more complicated, and the tests seem to fail with an older List::Util anyway.

yanick commented:
[typo] In all probabilty, it's now fixed.

[minimum versions] Yeah, I think it's be fair to set in dist.ini a reasonable minimum for all 3 List::* distros. 

And to be properly belt-and-suspenderish, I've added code that, when we `use ':all'`, check for which versions we have, exactly, and draw the list of functions in consequence.

yanick commented:
As far as for the tests that are still failing... the tidy rules for the xtest seem to be different than 'tidyall'? And  the documentation of List::UtilsBy don't mention some of its modules, so the POD coverage is complaining about that.  Do you want me to add documentation for thoses? (say the word, and I totally would)

autarch commented:
It'd probably be better to submit docs upstream for UtilsBy and add the missing ones based on that patch. Otherwise it'll be a pain to deal with in the future, as what I've done historically is just open the relevant module to copy & paste its docs.

autarch commented:
Here's another problem I just thought of. With the current version, it adapts if you install a new `List::Util` with more functions. This new version will require a new AllUtils release every time one of its three underlying modules adds something. That doesn't seem ideal.

yanick commented:
Yeah, you are right. With this,L::AU has to keep updated.

Hmmm... I just submitted a patch that adds some tsneaky tricks:

* If specific functions are requested and we know about it, we only use the related modules. Or we warn if the local modules are not up to snuff. This is good.
* If the functions are not known by L::AU, we `require` and peek in all sources modules, one by one, until we find the function (or run out of places to look for it). Worst case, we loaded all the modules, which is not worse than current implementation. 

* If `:all` is used, we go on and use the @EXPORT_OK of all source modules.

* If `List::AllUtils::froob`, where `froob` is a function we don't already know about, is called... well, we are screwed. For that to work, the user has to do `use List::AllUtils qw/ :all /;`.  But then, if the dependency versions of L::U, L::SU and L::UB are kept fresh, this shouldn't be too much of a pain. 

I guess the question is, are the savings of only loading the modules we really need worth the additional `:all` tags users will have to put if they want to use the full namespacey function names...

autarch commented:
```
autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ time perl -e1

real	0m0.003s
user	0m0.000s
sys	0m0.002s
autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ time perl -MList::Util -e1

real	0m0.009s
user	0m0.006s
sys	0m0.003s
autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ time perl -MList::Util -MList::SomeUtils -e1

real	0m0.019s
user	0m0.016s
sys	0m0.004s
autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ time perl -MList::Util -MList::SomeUtils -MList::UtilsBy -e1

real	0m0.024s
user	0m0.020s
sys	0m0.004s
```

So preloading all three modules takes a whole .02 seconds.

```
autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ perl -E 'system(q{smem}, q{-P}, q{perl});'
  PID User     Command                         Swap      USS      PSS      RSS 
17387 autarch  perl -E system(q{smem}, q{-        0     1468     1471     2200 

autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ perl -MList::Util -E 'system(q{smem}, q{-P}, q{perl});'
  PID User     Command                         Swap      USS      PSS      RSS 
17476 autarch  perl -MList::Util -E system        0     2092     2096     2888 

autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ perl -MList::Util -MList::SomeUtils -E 'system(q{smem}, q{-P}, q{perl});'
  PID User     Command                         Swap      USS      PSS      RSS 
17564 autarch  perl -MList::Util -MList::S        0     3636     3640     4436 

autarch@houseabsolute:~/projects/List-AllUtils (yanick-list-utilsby-with-new-exports *)$ perl -MList::Util -MList::SomeUtils -MList::UtilsBy -E 'system(q{smem}, q{-P}, q{perl});'
  PID User     Command                         Swap      USS      PSS      RSS 
17651 autarch  perl -MList::Util -MList::S        0     3760     3764     4560 
```

And loading all of them uses about 2MB of memory.

Overall, I think it'd be best to load them all up front and keep it simple.

yanick commented:
Sounds reasonable. And if there is ever a need, we can always pull back this branch from the mothballs.

Thanks!

autarch commented:
Closing since I merged the other PR that just loads all 3 modules without the clever bits.

yanick commented:
\o/ Thanks!
